cpmpy/solvers/exact.py

```diff
@@ -0,0 +1,492 @@
+        self.xct_solver.setOption("inp-dombreaklim", "0") # no dominance breaking to preserve solutions
+
+        # for solving with assumption variables,
+        self.assumption_dict = None
+
+        # objective can only be set once, so keep track of this
+        self.solver_is_initialized = False
+        self.has_objective = False
+        self.objective_minimize = True
+
+        # initialise everything else and post the constraints/objective
+        super().__init__(name="exact", cpm_model=cpm_model)
+
+    def getSolAndObj(self):
+        if not self.xct_solver.hasSolution():
+            self.objective_value_ = None
+            return False
+
+        # fill in variable values
+        lst_vars = list(self.user_vars)
+        exact_vals = self.xct_solver.getLastSolutionFor(self.solver_vars(lst_vars))
+        for cpm_var, val in zip(lst_vars,exact_vals):
+            cpm_var._value = bool(val) if isinstance(cpm_var, _BoolVarImpl) else val # xct value is always an int
+
+        # translate objective
+        self.objective_value_ = self.xct_solver.getObjectiveBounds()[1] # last upper bound to the objective
+        if not self.objective_minimize:
+            self.objective_value_ = -self.objective_value_
+
+        return True
+
+    def solve(self, time_limit=None, assumptions=None, **kwargs):
+        # TODO: test this function
+        """
+            Call Exact
+
+            Arguments:
+            - time_limit:  maximum solve time in seconds (float, optional)
+            - assumptions: list of CPMpy Boolean variables (or their negation) that are assumed to be true.
+                           For repeated solving, and/or for use with s.get_core(): if the model is UNSAT,
+                           get_core() returns a small subset of assumption variables that are unsat together.
+
+            Additional keyword arguments:
+            The Exact solver parameters are defined by https://gitlab.com/JoD/exact/-/blob/main/src/Options.hpp#L207
+        """
+        from exact import Exact as xct
+
+        if not self.solver_is_initialized:
+            assert not self.has_objective
+            self.xct_solver.init([],[])
+            self.solver_is_initialized=True
+
+        # set additional keyword arguments
+        for (kw, val) in kwargs.items():
+            self.xct_solver.setOption(kw,str(val))
+
+        # set assumptions
+        if assumptions is not None:
+            assert(all(is_bool(v) for v in assumptions))
+            assump_vals = [int(not isinstance(v, NegBoolView)) for v in assumptions]
+            assump_vars = [self.solver_var_no_num(v._bv if isinstance(v, NegBoolView) else v) for v in assumptions]
+            self.assumption_dict = {xct_var: (xct_val,cpm_assump) for (xct_var, xct_val, cpm_assump) in zip(assump_vars,assump_vals,assumptions)}
+            self.xct_solver.setAssumptions(assump_vars, assump_vals);
+            # NOTE: setAssumptions does not clear previous assumptions, is this ok?
```

Comment:
good point... I kind-of expect only the vars in the latest call should be True. At the very least the function definition should document what it does (e.g. only latest, or all)
in OrTools we first clear, and then add
(how can you actually clear them from the CPMpy interface? we can only give those that should be true... if we can't clear from the interface, I think we should clear by default)

Reply:
Yeah, cleared by default by `solve` and `solveAll`, seems best for now.